File: boxnote2md/render/blocks.py
"""ブロックレベルノードの Markdown 化。"""

from __future__ import annotations

from boxnote2md.render.document import (
    BLOCK_RENDERERS,
    Node,
    RenderContext,
    render_block,
    render_children_inline,
)


def render_paragraph(node: Node, ctx: RenderContext) -> str:
    inner = render_children_inline(node, ctx)
    return inner  # 段落間の \n\n は document.py 側で結合


def render_heading(node: Node, ctx: RenderContext) -> str:
    level = int(node.get("attrs", {}).get("level", 1) or 1)
    level = max(1, min(level, 6))
    inner = render_children_inline(node, ctx).strip()
    return f"{'#' * level} {inner}"


def render_horizontal_rule(_node: Node, _ctx: RenderContext) -> str:
    return "---"


def _render_children_blocks(node: Node, ctx: RenderContext) -> list[str]:
    """ブロック子要素を順にレンダリングし、文字列リストで返す。"""
    parts: list[str] = []
    for child in node.get("content", []) or []:
        ct = child.get("type")
        if ct == "paragraph":
            parts.append(render_children_inline(child, ctx))
        elif ct in BLOCK_RENDERERS:
            parts.append(render_block(child, ctx))
        else:
            ctx.warn(f"unknown child of {node.get('type')!r}: {ct!r}")
            parts.append(render_children_inline(child, ctx))
    return [p for p in parts if p]


def render_blockquote(node: Node, ctx: RenderContext) -> str:
    blocks = _render_children_blocks(node, ctx)
    body = "\n\n".join(blocks)
    return "\n".join(f"> {ln}" if ln else ">" for ln in body.split("\n"))


def render_code_block(node: Node, ctx: RenderContext) -> str:
    language = (node.get("attrs", {}).get("language") or "").strip()
    # text ノードを連結 (marks は無視; コード内の装飾は MD では表現できない)
    text_parts: list[str] = []
    for child in node.get("content", []) or []:
        if child.get("type") == "text":
            text_parts.append(child.get("text", ""))
    code = "".join(text_parts)
    # 言語名は Markdown 慣習で小文字化
    fence_lang = language.lower() if language else ""
    return f"```{fence_lang}\n{code}\n```"


def render_call_out_box(node: Node, ctx: RenderContext) -> str:
    """call_out_box は MD に対応物が無いので blockquote に退避。

    emoji があれば先頭段落に付与。--keep-styles 時は HTML <div> に背景色付き。
    """
    attrs = node.get("attrs", {}) or {}
    emoji = attrs.get("emoji") or ""
    bg = attrs.get("backgroundColor")

    blocks = _render_children_blocks(node, ctx)
    if not blocks:
        return ""

    if emoji:
        blocks[0] = f"{emoji} {blocks[0]}"

    if ctx.keep_styles and bg:
        body = "\n\n".join(blocks)
        return f'<div style="background-color:{bg}">\n\n{body}\n\n</div>'

    body = "\n\n".join(blocks)
    return "\n".join(f"> {ln}" if ln else ">" for ln in body.split("\n"))


RENDERERS = {
    "paragraph": render_paragraph,
    "heading": render_heading,
    "horizontal_rule": render_horizontal_rule,
    "blockquote": render_blockquote,
    "code_block": render_code_block,
    "call_out_box": render_call_out_box,
}

File: internal/render/blocks.go
package render

import (
	"fmt"
	"strings"

	"github.com/nanakanok/boxnote2md-cli/internal/boxnote"
)

func renderParagraph(n *boxnote.Node, ctx *Context) string {
	return renderChildrenInline(n, ctx)
}

func renderHeading(n *boxnote.Node, ctx *Context) string {
	level := n.AttrInt("level", 1)
	if level < 1 {
		level = 1
	}
	if level > 6 {
		level = 6
	}
	inner := strings.TrimSpace(renderChildrenInline(n, ctx))
	return strings.Repeat("#", level) + " " + inner
}

func renderBlockquote(n *boxnote.Node, ctx *Context) string {
	blocks := renderChildrenBlocks(n, ctx)
	body := strings.Join(blocks, "\n\n")
	return prefixLines(body, "> ", ">")
}

func renderCodeBlock(n *boxnote.Node, ctx *Context) string {
	language := strings.ToLower(strings.TrimSpace(n.AttrString("language")))
	var sb strings.Builder
	for _, c := range n.Content {
		if c.Type == "text" {
			sb.WriteString(c.Text)
		}
	}
	return fmt.Sprintf("```%s\n%s\n```", language, sb.String())
}

func renderCallOutBox(n *boxnote.Node, ctx *Context) string {
	emoji := n.AttrString("emoji")
	bg := n.AttrString("backgroundColor")
	blocks := renderChildrenBlocks(n, ctx)
	if len(blocks) == 0 {
		return ""
	}
	if emoji != "" {
		blocks[0] = emoji + " " + blocks[0]
	}
	if ctx.KeepStyles && bg != "" {
		body := strings.Join(blocks, "\n\n")
		return fmt.Sprintf(`<div style="background-color:%s">

%s

</div>`, bg, body)
	}
	body := strings.Join(blocks, "\n\n")
	return prefixLines(body, "> ", ">")
}

// prefixLines は文字列の各行に prefix (空行には emptyPrefix) を付与する。
func prefixLines(s, prefix, emptyPrefix string) string {
	lines := strings.Split(s, "\n")
	for i, ln := range lines {
		if ln == "" {
			lines[i] = emptyPrefix
		} else {
			lines[i] = prefix + ln
		}
	}
	return strings.Join(lines, "\n")
}
